server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotShutdownIT.java

```diff
@@ -0,0 +1,518 @@
+            final var entriesForRepo = SnapshotsInProgress.get(state).forRepo(repoName);
+            if (entriesForRepo.isEmpty()) {
+                return true;
+            }
+            assertThat(entriesForRepo, hasSize(1));
+            final var shardSnapshotStatuses = entriesForRepo.iterator().next().shards().values();
+            assertThat(shardSnapshotStatuses, hasSize(1));
+            assertThat(
+                shardSnapshotStatuses.iterator().next().state(),
+                oneOf(SnapshotsInProgress.ShardState.INIT, SnapshotsInProgress.ShardState.SUCCESS)
+            );
+            return false;
+        });
+
+        PlainActionFuture.<Void, RuntimeException>get(
+            fut -> putShutdownMetadata(
+                clusterService,
+                SingleNodeShutdownMetadata.builder()
+                    .setType(SingleNodeShutdownMetadata.Type.RESTART)
+                    .setStartedAtMillis(clusterService.threadPool().absoluteTimeInMillis())
+                    .setReason("test"),
+                originalNode,
+                fut
+            ),
+            10,
+            TimeUnit.SECONDS
+        );
+        assertFalse(snapshotCompletesWithoutPausingListener.isDone());
+        unblockAllDataNodes(repoName); // lets the shard snapshot continue so the snapshot can succeed
+        assertEquals(SnapshotState.SUCCESS, snapshotFuture.get(10, TimeUnit.SECONDS).getSnapshotInfo().state());
+        safeAwait(snapshotCompletesWithoutPausingListener);
+        clearShutdownMetadata(clusterService);
+    }
+
+    public void testRemoveNodeDuringSnapshot() throws Exception {
+        internalCluster().ensureAtLeastNumDataNodes(1);
+        final var originalNode = internalCluster().startDataOnlyNode();
+        final var indexName = randomIdentifier();
+        createIndexWithContent(indexName, indexSettings(1, 0).put(REQUIRE_NODE_NAME_SETTING, originalNode).build());
+
+        final var repoName = randomIdentifier();
+        createRepository(repoName, "mock");
+
+        final var clusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        final var snapshotFuture = startFullSnapshotBlockedOnDataNode(randomIdentifier(), repoName, originalNode);
+        final var snapshotPausedListener = createSnapshotPausedListener(clusterService, repoName, indexName);
+
+        updateIndexSettings(Settings.builder().putNull(REQUIRE_NODE_NAME_SETTING), indexName);
+        putShutdownForRemovalMetadata(originalNode, clusterService);
+        unblockAllDataNodes(repoName); // lets the shard snapshot abort, which frees up the shard so it can move
+        safeAwait(snapshotPausedListener);
+
+        // snapshot completes when the node vacates even though it hasn't been removed yet
+        assertEquals(SnapshotState.SUCCESS, snapshotFuture.get(10, TimeUnit.SECONDS).getSnapshotInfo().state());
+
+        if (randomBoolean()) {
+            internalCluster().stopNode(originalNode);
```

Comment:
Can we check that the cluster is still green after the nodes is stopped? That would verify that the snapshot did not complete until node was vacated.

Reply:
We _can_ (see ae572337e840db0a2b8545344a68e8b406fec4f1) but I admit I don't exactly follow how it verifies what you claim it does.

Reply:
With `ensureGreen`, I also do not think it validates that. What I wanted to suggest was to just check that the cluster is green with no wait, something like:

```
assertThat(clusterAdmin().prepareHealth().get().getStatus(), is(ClusterHealthStatus.GREEN));
```

That would validate that the snapshot waited for the node to vacate, since stopping the node leads to no unassigned shards?

Reply:
I see, although I don't think the snapshot could have reached state `SUCCESS` in that case. But I believe we can indeed assert that we never see any unassigned shards in these tests, see https://github.com/elastic/elasticsearch/pull/103905.